modules/configuration/src/main/java/org/apache/ignite/internal/configuration/hocon/HoconPrimitiveConfigurationSource.java

```diff
@@ -0,0 +1,171 @@
+        return new IllegalArgumentException(format(
+            "'%s' is expected as a type for the '%s' configuration value",
+            unbox(clazz).getSimpleName(), formatArrayPath(path, idx)
+        ));
+    }
+
+    /**
+     * Non-null wrapper over {@link TypeUtils#unboxed}.
+     */
+    private static Class<?> unbox(Class<?> clazz) {
+        assert !clazz.isPrimitive();
+
+        Class<?> unboxed = TypeUtils.unboxed(clazz);
+
+        return unboxed == null ? clazz : unboxed;
+    }
+
+    /**
+     * Extracts the value from the given {@link ConfigValue} based on the expected type.
+     */
+    public static <T> T unwrapPrimitive(List<String> path, ConfigValue hoconCfgValue, Class<T> clazz, int idx) {
+        assert !clazz.isArray();
+        assert !clazz.isPrimitive();
+
+        if (clazz == String.class) {
+            if (hoconCfgValue.valueType() != STRING)
+                throw wrongTypeException(path, clazz, idx);
+
+            return clazz.cast(hoconCfgValue.unwrapped());
+        }
+        else if (clazz == Boolean.class) {
+            if (hoconCfgValue.valueType() != BOOLEAN)
+                throw wrongTypeException(path, clazz, idx);
+
+            return clazz.cast(hoconCfgValue.unwrapped());
+        }
+        else if (clazz == Character.class) {
+            if (hoconCfgValue.valueType() != STRING || hoconCfgValue.unwrapped().toString().length() != 1)
+                throw wrongTypeException(path, clazz, idx);
+
+            return clazz.cast(hoconCfgValue.unwrapped().toString().charAt(0));
+        }
+        else if (Number.class.isAssignableFrom(clazz)) {
+            if (hoconCfgValue.valueType() != NUMBER)
+                throw wrongTypeException(path, clazz, idx);
+
+            Number numberValue = (Number)hoconCfgValue.unwrapped();
+
+            if (clazz == Byte.class) {
+                checkBounds(path, numberValue, Byte.MIN_VALUE, Byte.MAX_VALUE, idx);
+
+                return clazz.cast(numberValue.byteValue());
+            }
+            else if (clazz == Short.class) {
+                checkBounds(path, numberValue, Short.MIN_VALUE, Short.MAX_VALUE, idx);
+
+                return clazz.cast(numberValue.shortValue());
+            }
+            else if (clazz == Integer.class) {
+                checkBounds(path, numberValue, Integer.MIN_VALUE, Integer.MAX_VALUE, idx);
+
+                return clazz.cast(numberValue.intValue());
+            }
+            else if (clazz == Long.class)
+                return clazz.cast(numberValue.longValue());
+            else if (clazz == Float.class)
+                return clazz.cast(numberValue.floatValue());
+            else if (clazz == Double.class)
+                return clazz.cast(numberValue.doubleValue());
+        }
+
+        throw new IllegalArgumentException("Unsupported type: " + clazz);
+    }
+
+    /**
+     * Checks that a number fits into the given bounds.
```

Comment:
Params and return are missing

Reply:
Yeah, they are missing on many methods here

Reply:
Done